Annotate Storage for nullability

roji opened the pull request:
Since @AndriySvyryd and @smitpatel seemed to enjoy reviewing previous nullability PRs so much, I thought I'd give @ajcvickers the same joy - there's a commit in here just for type mapping (including converters) which can be reviewed separately.

Includes includes some leftovers from ~#23381 and~ #23376

/cc @smitpatel for some minor query things

roji commented on src/EFCore.Relational/Query/Internal/BufferedDataReader.cs:
```diff
@@ -1270,6 +1270,8 @@ private void InitializeFields()
                     for (var i = 0; i < _columns.Count; i++)
                     {
                         var column = _columns[i];
+                        // TODO-NULLABLE: Interestingly, there doesn't seem to be a single place in the codebase where we instantiate
```
/cc @smitpatel

smitpatel replied:
What do you mean? We create ReaderColumns!

roji replied:
Sorry, somehow missed that creation happens via the Create factory method.

So there's an assumption here that if the first column has a name (checked above), then all columns have a name, right?

smitpatel replied:
NCND truly.

smitpatel replied:
At least update the comment. :trollface: 
Put bangs if necessary. If you feel we should investigate, file an issue and assign to me. I will look into it.

roji replied:
Not really feeling we should investigate - was mainly curious about this assumption. You can investigate if you want to.

roji commented on src/EFCore.Relational/Query/Internal/FromSqlParameterExpandingExpressionVisitor.cs:
```diff
@@ -128,8 +128,9 @@ public class FromSqlParameterExpandingExpressionVisitor : ExpressionVisitor
                                 }
                             }
 
+                            // TODO-NULLABLE: we always set the name on ParameterExpression?
```
/cc @smitpatel

smitpatel replied:
Yes, parameter names in the query always have names

roji commented on src/EFCore.Relational/Query/SqlExpressionFactory.cs:
```diff
@@ -34,7 +34,9 @@ public SqlExpressionFactory([NotNull] SqlExpressionFactoryDependencies dependenc
 
             _typeMappingSource = dependencies.TypeMappingSource;
-            _boolTypeMapping = _typeMappingSource.FindMapping(typeof(bool));
+            // TODO-NULLABLE: we assume all relational providers have a mapping for bool
```
/cc @smitpatel

smitpatel replied:
Some mappings are required. :trollface: Though it does not have to map to a bool on server, can be mapped using a value converter too, we don't care.
We should bang this and perhaps add a validation code somewhere in type mapper
@ajcvickers ?

smitpatel replied:
Discussed with Arthur about this. This mapping is crucial for query to work even for simple predicate kind of scenario (since they are of bool type). By default providers don't really need bool in database. They just need some support for bool even through value conversion. Also built-in we provide bool to string & bool to int conversion in the box. So it should just work in 99.99% cases. It may cause a failure for an ORM written by amoeba targeting some database but if neither of that possible than it would be probably write only database.

roji commented on src/EFCore.Relational/Storage/Internal/RawRelationalParameter.cs:
```diff
@@ -71,6 +64,8 @@ public override void AddDbParameter(DbCommand command, object value)
                 value = cloneable.Clone();
             }
 
+            // TODO-NULLABLE: This looks like a bug - if value isn't a a DbParameter, we attempt to add a raw value,
```
/cc @AndriySvyryd @smitpatel

smitpatel replied:
This will always be a DbParameter.

roji replied:
Thanks, the code currently seems written to allow otherwise, will change to an assertion

roji commented on src/EFCore.Relational/Storage/RelationalConnection.cs:
```diff
@@ -79,14 +82,15 @@ protected RelationalConnection([NotNull] RelationalConnectionDependencies depend
         /// </summary>
         public virtual Guid ConnectionId { get; } = Guid.NewGuid();
 
+        // TODO-NULLABLE: Can this actually return null (see comment)?
```
/cc @AndriySvyryd

AndriySvyryd replied:
Probably not

roji replied:
Removing comment.

roji commented on src/EFCore.Relational/Storage/RelationalConnection.cs:
```diff
@@ -237,7 +242,8 @@ public virtual void EnlistTransaction(Transaction transaction)
         /// </summary>
         /// <param name="transaction"> The transaction to be used. </param>
-        protected virtual void ConnectionEnlistTransaction([NotNull] Transaction transaction)
+        // TODO-NULLABLE: This gets called with null, and DbConnection.EnlistTransaction does accept null
```
/cc @AndriySvyryd

AndriySvyryd replied:
So what's the issue?

roji replied:
Just mentioning the annotation change from `[NotNull]` to `[CanBeNull]`.

roji commented on src/EFCore.Relational/Storage/RelationalConnection.cs:
```diff
@@ -327,6 +329,7 @@ public virtual IDbContextTransaction BeginTransaction(IsolationLevel isolationLe
         /// </summary>
         /// <param name="isolationLevel"> The isolation level to use for the transaction. </param>
         /// <returns> The newly created transaction. </returns>
+        // TODO-NULLABLE: Spelling mistake :(
```
@AndriySvyryd obsolete and fix typo? Spelling errors are worse than data corruption

AndriySvyryd replied:
Just fix it, this is new in 6.0

roji replied:
OK, perfect.

roji commented on src/EFCore.Relational/Storage/RelationalConnection.cs:
```diff
@@ -428,7 +432,8 @@ public virtual IDbContextTransaction UseTransaction(DbTransaction transaction)
         /// <param name="transactionId"> The unique identifier for the transaction. </param>
         /// <returns> An instance of <see cref="IDbTransaction" /> that wraps the provided transaction. </returns>
-        public virtual IDbContextTransaction UseTransaction(DbTransaction transaction, Guid transactionId)
+        // TODO-NULLABLE: Note that this can return null (if correct, need to update doc comment)
```
/cc @AndriySvyryd

AndriySvyryd replied:
Yes, doc should be fixed

roji replied:
Documented null return, and also corrected IDbTransaction->IDbContextTransaction.

Also annotated the method with [NotNullIfNotNull], since null only gets returned for a null parameter.

roji commented on src/EFCore.Relational/Storage/RelationalTypeMapping.cs:
```diff
@@ -277,6 +279,8 @@ protected RelationalTypeMapping(RelationalTypeMappingParameters parameters)
 
             var storeType = parameters.StoreType;
 
+            // TODO-NULLABLE: It seems like RelationalTypeMappingParameters.StoreType is non-nullable, so this check is redundant?
```
/cc @ajcvickers

AndriySvyryd replied:
If only there had been a way to indicate that it is never `null` :trollface:

roji replied:
So... OK to remove this check?

roji commented on src/EFCore.Relational/Storage/TypeMaterializationInfo.cs:
```diff
@@ -65,12 +67,13 @@ public class TypeMaterializationInfo
-            [CanBeNull] IProperty property,
-            [NotNull] RelationalTypeMapping mapping,
+            [CanBeNull] IProperty? property,
+            [CanBeNull] RelationalTypeMapping? mapping,
```
/cc @smitpatel

smitpatel replied:
Query does not use this code path at all. But TypeMapping should be required.

roji replied:
Ah I see now - the mapping is optional because of the obsoleted constructors which don't accept a type mapping (or try to infer it) - these ctors are never used in our codebase. If type mapping is actually null, TypedRelationalValueBufferFactoryFactory would fail.

Am removing these obsoleted constructors for now - @AndriySvyryd can you please confirm?

roji commented on src/EFCore.Relational/Storage/TypedRelationalValueBufferFactoryFactory.cs:
```diff
@@ -193,7 +194,8 @@ public virtual IReadOnlyList<Expression> CreateAssignmentExpressions([NotNull] I
             bool box = true)
         {
-            var getMethod = materializationInfo.Mapping.GetDataReaderMethod();
+            // TODO-NULLABLE: Smit to confirm
```
/cc @smitpatel

smitpatel replied:
TypeMapping is required component in materializationInfo.

roji replied:
Continued in https://github.com/dotnet/efcore/pull/23423#discussion_r528173909

roji commented on src/EFCore/Infrastructure/ExpressionExtensions.cs:
```diff
@@ -326,7 +326,8 @@ public static IReadOnlyList<MemberInfo> GetMemberAccessList([NotNull] this Lambd
             IPropertyBase property)
 #pragma warning restore IDE0060 // Remove unused parameter
-            => (TValue)valueBuffer[index];
+        // TODO-NULLABLE: if the buffer contains null but TValue is non-nullable, are we expected to throw here?
```
/cc @smitpatel

smitpatel replied:
yes, we expected to throw. This is InMemory version of trying to call GetX method on a null column value. :trollface:

roji replied:
:+1:

roji commented on src/EFCore/Metadata/MemberIdentity.cs:
```diff
@@ -78,9 +79,12 @@ public static MemberIdentity Create([CanBeNull] MemberInfo? memberInfo)
+        public string Name
         {
-            [DebuggerStepThrough] get => MemberInfo?.GetSimpleMemberName() ?? (string?)_nameOrMember;
+            [DebuggerStepThrough] get => MemberInfo?.GetSimpleMemberName()
```
Fixup for #23376

roji replied:
@AndriySvyryd @smitpatel making Name throw for null/empty fails test Non_nullable_property_through_optional_navigation, for example: [this line](https://github.com/dotnet/efcore/blob/main/src/EFCore/Metadata/Internal/InternalForeignKeyBuilder.cs#L408) causes a navigation to be created with a null MemberIdentity, and its name is accessed.

Should we make MemberIdentity.Name nullable again and introduce IsNone checks where relevant ([option 2](https://github.com/dotnet/efcore/pull/23381#discussion_r527935981)), or do you think we need a deeper change here?

AndriySvyryd replied:
No, it should pass `MemberIdentity.None` when name is `null`

roji replied:
Am not sure I understand... If we make MemberIdentity.Name to throw when the MemberIdentity is None, there are many call sites where we'd need to add an IsNone check before accessing Name to avoid it being thrown (e.g. all the places where Name is accessed and then compared to null in metadata). Can you confirm this is what you want?

roji replied:
Am backing this out for now - this is a non-trivial change I'll do separately (made a note in #19007.

roji commented on src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.ExpressionVisitors.cs:
```diff
@@ -124,17 +124,15 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                     var navigation = memberIdentity.MemberInfo != null
                         ? entityType.FindNavigation(memberIdentity.MemberInfo)
-                        : entityType.FindNavigation(memberIdentity.Name!);
+                        : entityType.FindNavigation(memberIdentity.Name);
```
Fixup for #23381

roji commented on src/EFCore/Storage/ExecutionStrategyDependencies.cs:
```diff
@@ -56,8 +58,8 @@ public sealed record ExecutionStrategyDependencies
-            [CanBeNull] IDbContextOptions options,
-            [CanBeNull] IDiagnosticsLogger<DbLoggerCategory.Infrastructure> logger)
+            [CanBeNull] IDbContextOptions? options,
+            [NotNull] IDiagnosticsLogger<DbLoggerCategory.Infrastructure> logger)
```
/cc @AndriySvyryd

AndriySvyryd replied:
Probably a leftover from a refactoring

AndriySvyryd replied:
Shouldn't `IDbContextOptions` be non-nullable too?

roji replied:
Have made this change.

roji commented on src/EFCore/Storage/ValueBuffer.cs:
```diff
@@ -65,6 +67,7 @@ public int Count
         ///     Gets a value indicating whether the value buffer is empty.
         /// </summary>
         public bool IsEmpty
+        // TODO-NULLABLE: We have an assertion in the ctor that this will never happen? What about _value.Length == 0?
```
/cc @smitpatel

smitpatel replied:
File a new issue, I will investigate.

roji replied:
#23428

roji commented on src/EFCore/Storage/ValueConversion/BytesToStringConverter.cs:
```diff
@@ -19,10 +21,11 @@ public class BytesToStringConverter : ValueConverter<byte[], string>
             : base(
-                v => v == null ? null : Convert.ToBase64String(v),
-                v => v == null ? null : Convert.FromBase64String(v),
+                // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
```
@ajcvickers discussed this situation offline, am leaving the code as is and adding bangs, until we revisit as part of #13850

roji replied:
#23428

smitpatel replied:
Wouldn't `null!` throw runtime error?

smitpatel replied:
And if that code path is not being triggered then is the null check here necessary?

roji commented on test/EFCore.Specification.Tests/ApiConsistencyTestBase.cs:
```diff
@@ -670,7 +670,13 @@ where ns.StartsWith("Microsoft.Entity", StringComparison.Ordinal)
                        && (it.GetConstructors().Length != 1
                            || it.GetConstructors()[0].GetParameters().Length == 0
-                           || it.GetConstructors()[0].GetParameters()[0].Name != "dependencies")
+                           || it.GetConstructors()[0].GetParameters()[0].Name != "dependencies"
```
Fixup to #23376

roji commented on src/EFCore.Relational/Storage/RelationalGeometryTypeMapping.cs:
```diff
@@ -83,9 +86,11 @@ public override DbParameter CreateParameter(DbCommand command, string name, obje
+            parameter.Value = value is null
                 ? DBNull.Value
-                : SpatialConverter.ConvertToProvider(value);
+                : SpatialConverter is null
```
/cc @bricelam on the one hand, we annotated SpatialConverter as [CanBeNull] in the ctor, but on the other we required it here (Npgsql, which doesn't require a converter, passes NullValueConverter currently). This PR changes to really not require a converter, or we could annotate SpatialConverter as required if we want.

smitpatel replied:
Also see https://github.com/dotnet/efcore/issues/21558

roji commented on src/EFCore/Storage/ExecutionStrategy.cs:
```diff
@@ -392,8 +394,10 @@ protected internal virtual bool ShouldVerifySuccessOn([NotNull] Exception except
-            [NotNull] Func<Exception, TResult> exceptionHandler)
+            [CanBeNull] Exception? exception,
+            [NotNull] Func<Exception?, TResult> exceptionHandler)
+            // TODO-NULLABLE: It's a bit weird to call the handler with a null exception. We could refactor this as UnwrapDbUpdateException,
```
/cc @AndriySvyryd

AndriySvyryd replied:
You mean that we should have an exclusion list for exceptions that we'll never retry?

roji replied:
Looking at this again, I think the current situation is fine, with the handler annotated to accept a nullable exception.

smitpatel commented on src/EFCore.Relational/Query/QuerySqlGenerator.cs:
```diff
@@ -303,6 +303,8 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
         protected override Expression VisitTableValuedFunction(TableValuedFunctionExpression tableValuedFunctionExpression)
         {
             Check.NotNull(tableValuedFunctionExpression, nameof(tableValuedFunctionExpression));
+            Check.DebugAssert(tableValuedFunctionExpression.Alias is not null,
```
Just bang these instead.

roji replied:
When it's the contract of a method about its parameters (like Check.NotNull), I prefer listing it explicitly as such in this way. But if you're strongly against it for some reason I change to bang.

smitpatel replied:
Can you explain how it is contract of a method? The reason I prefer to bang this is that it is only used in one place and inside this method body only. If putting the DebugAssert would have helped in multiple places or in multiple code path then there is some value to it, otherwise I am leaning towards of just putting bang here.

roji replied:
> Can you explain how it is contract of a method?

Well, just like VisitTableValuedFunction requires that tableValuedFunctionExpression be non-null (hence Check.NotNull), it also requires tableValuedFunctionExpression.Alias to be non-null - the two are API contracts of the method (if you violate either, the method will not work). Putting this up-front like this is a form of documentation, code contract etc.

Having said that, there is no runtime difference between either... if you still prefer a bang I will do that.

smitpatel replied:
If you want to specify the contract to be "tableValuedFunctionExpression.Alias" is not-null then that contract should live on the ctor of TableValueFunctionExpression itself. This API has contractual agreement towards only not passing a null tableValuedFunctionExpression here. For any component of the any expression, this is just consumer API not binder.

roji replied:
> If you want to specify the contract to be "tableValuedFunctionExpression.Alias" is not-null then that contract should live on the ctor of TableValueFunctionExpression itself.

If TableValueFunctionExpression.Alias should never ever have null, than you would be right (and we'd make Alias non-nullable) - but that's not the case, right? It seems legal for Alias to be null in some situations (i.e. before the alias is calculated).

So the contract isn't on the TableValueFunctionExpression type - it's on VisitTableValuedFunction, specifying the contents of TableValueFunctionExpression which are legal for its specific invocation.

smitpatel replied:
> If TableValueFunctionExpression.Alias should never ever have null, than you would be right (and we'd make Alias non-nullable) - but that's not the case, right?

Incorrect. TableValuedFunctionExpression.Alias is **always** non-null. We always calculate alias for the locations required. We only uniquify them at later stage. We _never_ assign alias which was null before.

roji replied:
That's great... In that case, we can just override TableValueFunctionExpression.Alias specifically to make it non-nullable. I just pushed a commit that does that (for TableValueFunctionExpression, FromSqlExpression, SetOperationBase).

smitpatel commented on src/EFCore.Relational/Query/SqlExpressions/FromSqlExpression.cs:
```diff
@@ -54,6 +55,16 @@ public FromSqlExpression([NotNull] string alias, [NotNull] string sql, [NotNull]
+        ///     The alias assigned to this table source.
+        /// </summary>
+        [CA.NotNull]
+        public override string? Alias
```
Also TableExpression

smitpatel commented on src/EFCore.Relational/Query/QuerySqlGenerator.cs:
```diff
@@ -563,6 +565,8 @@ protected override Expression VisitSqlConstant(SqlConstantExpression sqlConstant
         protected override Expression VisitSqlParameter(SqlParameterExpression sqlParameterExpression)
         {
             Check.NotNull(sqlParameterExpression, nameof(sqlParameterExpression));
+            Check.DebugAssert(sqlParameterExpression.TypeMapping is not null,
```
Remove these 2 too.
TypeMapping is never null after translation.
ColumnExpression always originates from a table which has alias.

roji replied:
>  TypeMapping is never null after translation.

That's the point behind this assertion: unlike with TableExpression.Alias which really is never null (so it can be annotated as such and no assertion needed), SqlParameterExpression.TypeMapping is null in some cases.

But as you seem to really hate assertions I'll remove and bang instead.

BTW if we want to, we can consider making SqlExpression.TypeMapping non-nullable, and provide a separate nullable property for the type mapping to be used only during translation. This would allow us to remove the bangs when accessing TypeMapping from everywhere.

smitpatel replied:
My issue is not with assertions in general. My issue is assertion in the wrong places just because it suits NRT. If the TypeMapping is null after translation in error then there are plenty of visitors which go through it and assuming it is going to be there. SqlGenerator is very much last phase of query pipeline. If you want to assert that TypeMapping is non-null after translation then this is certainly not the place to assert it.
So please don't put assert in wrong places in query.

smitpatel replied:
cc: @AndriySvyryd

roji replied:
I don't have super strong feelings here... But given that TypeMapping is nullable (because used during translation), any visitor after translation accessing it is *assuming* it will be non-null, which is fine. And I tend to codify/document coding assumptions via assertions - it's logically the exact same thing as a bang, just more explicit. Assertions also seem particularly well-suited to me when they codify assumptions on method parameters (exactly like Check.NotNull), and they concentrate the assumptions at the beginning rather than having bangs interspersed later in the method code.

smitpatel replied:
Codify assumptions but not in wrong place.

AndriySvyryd commented on src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs:
```diff
@@ -91,17 +94,16 @@ protected override CoreTypeMapping FindMapping(in TypeMappingInfo mappingInfo)
             if (principals != null)
             {
-                for (var i = 0; i < principals.Count; i++)
+                foreach (var principal in principals)
```
Put that `for` back please, allocations matter

roji replied:
Sorry, missed that this was an IReadOnlyList

smitpatel commented on src/EFCore.Relational/Query/QuerySqlGenerator.cs:
```diff
@@ -329,6 +329,8 @@ protected override Expression VisitTableValuedFunction(TableValuedFunctionExpres
         protected override Expression VisitColumn(ColumnExpression columnExpression)
         {
             Check.NotNull(columnExpression, nameof(columnExpression));
+            Check.DebugAssert(columnExpression.Table.Alias is not null,
```
Remove this assert.

AndriySvyryd commented on src/EFCore.Relational/Storage/RelationalConnection.cs:
```diff
@@ -382,7 +382,8 @@ protected virtual DbTransaction ConnectionBeginTransation(IsolationLevel isolati
         /// <param name="cancellationToken"> A <see cref="CancellationToken" /> to observe while waiting for the task to complete. </param>
         /// <returns> The newly created transaction. </returns>
-        protected virtual ValueTask<DbTransaction> ConnectionBeginTransationAsync(
+        // TODO-NULLABLE: Spelling mistake :(
```
Can remove now

smitpatel commented on src/EFCore.Relational/Storage/IRelationalConnection.cs:
```diff
@@ -47,6 +50,7 @@ string ConnectionString
         ///         Note that the connection must be disposed by application code since it was not created by Entity Framework.
         ///     </para>
         /// </summary>
+        [CA.AllowNull]
```
Any reason this does not return nullable?

roji replied:
The implementation in RelationalConnection creates a new connection and returns that if there isn't one, so I'm assuming the contract is for IRelationalConnection.Connection to always return non-null. @AndriySvyryd?

AndriySvyryd replied:
Yes, this is necessary to make `DatabaseFacade.GetDbConnection()` never return `null`

smitpatel commented on src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs:
```diff
@@ -172,7 +174,7 @@ private static void FormatParameterValue(StringBuilder builder, object parameter
                     var isNullProperty = parameterValue.GetType().GetRuntimeProperty("IsNull");
                     if (isNullProperty != null
-                        && (bool)isNullProperty.GetValue(parameterValue))
+                        && isNullProperty.GetValue(parameterValue) is bool)
```
You also need to verify the bool value to be true to keep the code doing the thing it was doing before.

roji replied:
Oops, thanks..

smitpatel commented on src/EFCore.Relational/Storage/Internal/DbParameterCollectionExtensions.cs:
```diff
@@ -172,7 +174,7 @@ private static void FormatParameterValue(StringBuilder builder, object parameter
                     var isNullProperty = parameterValue.GetType().GetRuntimeProperty("IsNull");
                     if (isNullProperty != null
-                        && (bool)isNullProperty.GetValue(parameterValue))
+                        && isNullProperty.GetValue(parameterValue) is bool isNull && isNull)
```
break line. :trollface:

smitpatel commented on src/EFCore.Relational/Storage/RelationalTypeMappingSource.cs:
```diff
@@ -33,8 +36,8 @@ namespace Microsoft.EntityFrameworkCore.Storage
     {
-        private readonly ConcurrentDictionary<(RelationalTypeMappingInfo, Type, ValueConverter), RelationalTypeMapping> _explicitMappings
-            = new ConcurrentDictionary<(RelationalTypeMappingInfo, Type, ValueConverter), RelationalTypeMapping>();
+        private readonly ConcurrentDictionary<(RelationalTypeMappingInfo, Type?, ValueConverter?), RelationalTypeMapping?> _explicitMappings
```
new()?

roji replied:
I stopped doing these systematically - the plan would be to just let Rider do it at the solution level afterwards...

smitpatel commented on src/EFCore.Relational/Storage/RelationalTypeMappingInfo.cs:
```diff
@@ -233,7 +235,7 @@ public bool IsKeyOrIndex
         ///     The CLR type in the model.
         /// </summary>
-        public Type ClrType
+        public Type? ClrType
```
Just curious. ClrType could be null?

roji replied:
Scaffolding no?

smitpatel commented on src/EFCore/Storage/TypeMappingInfo.cs:
```diff
@@ -54,14 +56,13 @@ public TypeMappingInfo([NotNull] IProperty property)
             int? scale = null;
             bool? isUnicode = null;
-            for (var i = 0; i < principals.Count; i++)
+            foreach (var principal in principals)
```
? ? ?

roji replied:
Yep, another one I did by mistake, reverting.

AndriySvyryd commented on src/EFCore/Storage/TypeMappingSource.cs:
```diff
@@ -42,17 +44,16 @@ protected TypeMappingSource([NotNull] TypeMappingSourceDependencies dependencies
             if (principals != null)
             {
-                for (var i = 0; i < principals.Count; i++)
+                foreach (var principal in principals)
```
☝️

roji replied:
Actually I put these in to test if you guys were paying attention.

smitpatel commented:
I am still unsure what is happening with value conversions
- We are saying that null is handled already when calling into expression then why do we need additional null check?
- Is null! is obviously wrong? How is it different than calling null.Method()? So what additional value the null check provided.

While the places has been marked with #13850 
- That issue enables that null value can be converted to non-null in database, basically skipping our first check that null value converts to null value.
- But that non-null converted value would need to come from user, no? By default we can just convert null to null in our built-in value conversions.
- Even if we decide to change our default conversions to identify more types, we cannot change the existing ones fully at all without causing a data break, since by default earlier we stored null values in database and now we would be storing altogether a different value.

roji commented:
@smitpatel re value conversions, I briefly discussed with @ajcvickers offline and we just decided to defer this for now until #13850 is tackled - so I did the minimum to suppress warnings etc., without worrying right now about what #13850 will do.

> We are saying that null is handled already when calling into expression then why do we need additional null check?

It does seem that currently we have double null-checking: once in the core value conversion infrastructure (e.g. SanitizeConverter), and once in each built-in value converter implementation we have.

> Is null! is obviously wrong? How is it different than calling null.Method()? So what additional value the null check provided.

null! isn't wrong in the same sense as null.Method(), since the latter always fails with an NRE, while the former simply preserves our current runtime behavior, exactly as it is (that was the idea behind what I currently did). But down the road we should annotate better depending on what happens with #13850.

roji commented:
@AndriySvyryd @smitpatel coming back to this after a while... I squashed and rebased over latest main, and pulled out the MemberIdentity stuff (fixup to #23381) which requires a bit more looking into (noted in #19007).
